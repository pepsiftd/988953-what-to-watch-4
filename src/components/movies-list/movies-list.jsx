import React from 'react';
import PropTypes from 'prop-types';

import {SmallMovieCard} from '@/components/small-movie-card/small-movie-card';
import {withVideo} from '@/hocs/with-video/with-video';

const SmallMovieCardWithVideo = withVideo(SmallMovieCard);

const VIDEO_SETTINGS = {
  isMute: true,
  width: `280`,
  height: `175`,
};

const MoviesList = ({movies, renderItem, setActiveItem, clearActiveItem}) => {
  return (
    <div className="catalog__movies-list">
      {
        movies.map((movie) => {
          const videoSettings = Object.assign({}, {
            src: movie.preview,
            poster: movie.imageSrc,
          }, VIDEO_SETTINGS);

          return renderItem(SmallMovieCardWithVideo, movie.id, {
            key: movie.id,
            movie,
            onMouseEnter: setActiveItem,
            onMouseLeave: clearActiveItem,

            videoSettings,
          });
        })
      }
    </div>
  );
};

MoviesList.propTypes = {
  movies: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    imageSrc: PropTypes.string.isRequired,
    movieLink: PropTypes.string.isRequired,
    preview: PropTypes.string.isRequired,
  }).isRequired).isRequired,
  renderItem: PropTypes.func.isRequired,
  setActiveItem: PropTypes.func.isRequired,
  clearActiveItem: PropTypes.func.isRequired,
};

export {MoviesList};
