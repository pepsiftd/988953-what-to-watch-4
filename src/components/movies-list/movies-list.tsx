import React from 'react';

import {SmallMovieCard} from '@/components/small-movie-card/small-movie-card';
import {withVideo} from '@/hocs/with-video/with-video';
import {SMALL_CARD_VIDEO_SETTINGS} from '@/const';

const SmallMovieCardWithVideo = withVideo(SmallMovieCard);

interface Props {
  movies: {
    id: number;
    title: string;
    imageSrc: string;
    preview: string;
  }[];
  renderItem: (Item: React.ReactElement, id: number | string, itemProps: {}) => React.ReactElement;
  setActiveItem: (id: number) => void;
  clearActiveItem: () => void;
  cardsShowing: number;
  renderShowMore: () => void;
};

const MoviesList: React.FunctionComponent<Props> = ({movies, renderItem, setActiveItem, clearActiveItem, cardsShowing, renderShowMore = () => {}}) => {
  const cardsShowingCount = cardsShowing ? cardsShowing : movies.length;
  return (
    <React.Fragment>
      <div className="catalog__movies-list">
        {
          movies.slice(0, cardsShowingCount).map((movie) => {
            const videoSettings = Object.assign({}, {
              src: movie.preview,
              poster: movie.imageSrc,
            }, SMALL_CARD_VIDEO_SETTINGS);

            return renderItem(SmallMovieCardWithVideo, movie.id, {
              key: movie.id,
              movie,
              onMouseEnter: setActiveItem,
              onMouseLeave: clearActiveItem,

              videoSettings,
            });
          })
        }
      </div>
      {renderShowMore()}
    </React.Fragment>
  );
};

export {MoviesList};
