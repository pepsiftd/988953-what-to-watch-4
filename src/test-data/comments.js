const reviews = [
  {
    id: 1,
    user: {
      id: 4,
      name: `Courtney Cox`,
    },
    rating: 5.1,
    comment: `I don't know what to say but I can't help posting`,
    date: new Date(`August 3, 2020 UTC`),
  },
  {
    id: 2,
    user: {
      id: 2,
      name: `Mockingbird`,
    },
    rating: 8.3,
    comment: `This is not a movie of a simple kind`,
    date: new Date(`July 15, 2020 UTC`),
  },
  {
    id: 3,
    user: {
      id: 154,
      name: `Pelagia`,
    },
    rating: 1.5,
    comment: `Super mega exquisitely boring`,
    date: new Date(`June 1, 2020 UTC`),
  }
];

export {reviews};
