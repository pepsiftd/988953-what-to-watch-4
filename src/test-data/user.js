const userInfo = {
  id: 0,
  email: `example@email.com`,
  name: `Sarah Connor`,
  avatar: `img/avatar.jpg`,
};

export {userInfo};
