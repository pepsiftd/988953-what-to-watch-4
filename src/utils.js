const getUniqueItems = (list) => {
  return list.filter((it, i, array) => {
    return array.indexOf(it) === i;
  });
};

const extend = (a, b) => {
  return Object.assign({}, a, b);
};

const getMoviesByGenre = (movies, genre) => {
  return genre === `All genres` ? movies : movies.filter((movie) => movie.genre === genre);
};

const getGenresFromMovies = (movies) => {
  return getUniqueItems(movies.map((movie) => movie.genre));
};

export {extend, getMoviesByGenre, getUniqueItems, getGenresFromMovies};
