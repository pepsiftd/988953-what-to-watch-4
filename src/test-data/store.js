import {NameSpace} from '@/reducer/name-space';
import {movies, promoMovie} from '@/test-data/movies';
import {userInfo} from '@/test-data/user';
import {reviews} from '@/test-data/comments';
import {ALL_GENRES_FILTER} from '@/const';

const store = {
  [NameSpace.DATA]: {
    movies,
    promoMovie,
    favoriteMovies: [],
    reviews,
  },
  [NameSpace.APP]: {
    currentGenre: ALL_GENRES_FILTER,
  },
  [NameSpace.USER]: {
    authorizationStatus: `AUTHORIZED`,
    authorizationError: 200,
    authorizationInfo: userInfo,
  },
};

export {store};
