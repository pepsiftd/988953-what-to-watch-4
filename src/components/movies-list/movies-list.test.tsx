import * as React from 'react';
import * as renderer from 'react-test-renderer';

import {MoviesList} from './movies-list';
import {movies} from '@/test-data/movies';
import {noop} from '@/utils';

it(`MoviesList should render correctly`, () => {
  const tree = renderer.create(
      <MoviesList
        movies={movies}
        renderItem={noop}
        setActiveItem={noop}
        clearActiveItem={noop}
        cardsShowing={8}
        renderShowMore={noop}
      />
  );

  expect(tree).toMatchSnapshot();
});
