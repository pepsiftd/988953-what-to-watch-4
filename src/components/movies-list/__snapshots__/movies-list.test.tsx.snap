// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList should render correctly 1`] = `
<div
  className="catalog__movies-list"
/>
`;
